### PR TITLE
chore(llmobs): projects support for dne

### PR DESCRIPTION
**Pull request by gary-huang**
adds support for projects
1. if a project name is provided in pull_dataset or create_dataset* methods, the dataset will be searched for or created under the specified project
2. otherwise, the default project is used, or the project name from the `LLMObs.enable` call or the env var is used

with this snippet
```
LLMObs.enable(api_key=os.getenv("DD_API_KEY"), app_key=os.getenv("DD_APPLICATION_KEY"),  project_name="Onboarding", ml_app="Onboarding-ML-App")

dataset_name = "weatheraus-2025090900011"
dataset = LLMObs.create_dataset_from_csv(csv_path="/Users/gary.huang/Downloads/weatherAUS.csv", dataset_name=dataset_name, project_name="big-data-gh" , input_data_columns=["Date", "input", "Evaporation", "Sunshine"], expected_output_columns=["RainTomorrow"])

print(dataset.as_dataframe())

print(dataset.url)

time.sleep(30)

try:
    pds = LLMObs.pull_dataset(dataset_name=dataset_name)
except Exception as e:
    print(e)

pds = LLMObs.pull_dataset(dataset_name=dataset_name, project_name="big-data-gh")
print(pds.as_dataframe())

print(pds.url)
```
we get the following output
```
python test-project.py
3.15.0.dev13+gcb1068a93
        input_data expected_output input_data                    
              Date    RainTomorrow   Sunshine Evaporation   input
0       2008-12-01              No         NA          NA  Albury
1       2008-12-02              No         NA          NA  Albury
2       2008-12-03              No         NA          NA  Albury
3       2008-12-04              No         NA          NA  Albury
4       2008-12-05              No         NA          NA  Albury
...            ...             ...        ...         ...     ...
145455  2017-06-21              No         NA          NA   Uluru
145456  2017-06-22              No         NA          NA   Uluru
145457  2017-06-23              No         NA          NA   Uluru
145458  2017-06-24              No         NA          NA   Uluru
145459  2017-06-25              NA         NA          NA   Uluru

[145460 rows x 5 columns]
https://app.datadoghq.com/llm/datasets/7f81325e-dce1-49c7-b6bf-c75947490891
Dataset 'weatheraus-202509090928' not found in project Onboarding
        input_data expected_output input_data                    
              Date    RainTomorrow   Sunshine Evaporation   input
0       2008-12-05              No         NA          NA  Albury
1       2008-12-08              No         NA          NA  Albury
2       2008-12-10              No         NA          NA  Albury
3       2008-12-09             Yes         NA          NA  Albury
4       2008-12-12             Yes         NA          NA  Albury
...            ...             ...        ...         ...     ...
145455  2016-08-25              No         NA          NA   Uluru
145456  2016-11-20              No         NA          NA   Uluru
145457  2016-12-26             Yes         NA          NA   Uluru
145458  2016-12-31              No         NA          NA   Uluru
145459  2017-04-29              No         NA          NA   Uluru

[145460 rows x 5 columns]
https://app.datadoghq.com/llm/datasets/7f81325e-dce1-49c7-b6bf-c75947490891
```

the first pull hits an exception because the dataset wasn't found in the default project from the enable call, and the 2nd pull succeeds, all as expected

an alternate approach would be to initialize default project in `enable`, but that messes with the tests everywhere that don't have cassettes 

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/llmobs/_llmobs.py [hunk 1]
```diff
@@ -654,15 +665,23 @@ def enable(
             )
 
     @classmethod
-    def pull_dataset(cls, name: str) -> Dataset:
-        ds = cls._instance._dne_client.dataset_get_with_records(name)
+    def pull_dataset(cls, dataset_name: str, project_name: Optional[str] = None) -> Dataset:
```
Comment (sabrenner) [inline ddtrace/llmobs/_llmobs.py:650]: just sanity checking - we're ok with breaking `name` &rarr; `dataset_name` since DnE isn't GA yet, right?

File: ddtrace/llmobs/_llmobs.py [hunk 2]
```diff
@@ -654,15 +665,23 @@ def enable(
...
+        ds = cls._instance._dne_client.dataset_get_with_records(
+            dataset_name, cls._project_name if project_name is None else project_name
```
Comment (sabrenner) [inline ddtrace/llmobs/_llmobs.py]:
```suggestion
            dataset_name, (project_name or cls._project_name)
```
mostly style nit although maybe not desired anyhow as this will also default to `cls._project_name` if `project_name` is any kind of falsy value (like empty string), so definitely fine to leave as-is 😄

File: ddtrace/llmobs/_writer.py [hunk 3]
```diff
@@ -534,7 +548,10 @@ def dataset_bulk_upload(self, dataset_id: str, records: List[DatasetRecord]):
             raise ValueError(f"Failed to upload dataset from file: {resp.status} {resp.get_json()}")
         logger.debug("successfully uploaded with code %d", resp.status)
 
-    def project_create_or_get(self, name: str) -> str:
+    def project_create_or_get(self, name: Optional[str] = None) -> str:
+        if name is None or name == "":
```
Comment (sabrenner) [inline ddtrace/llmobs/_writer.py]:
```suggestion
        if not name:
```

File: ddtrace/llmobs/_writer.py [hunk 4]
```diff
@@ -362,23 +364,31 @@ def dataset_delete(self, dataset_id: str) -> None:
...
         if resp.status != 200:
-            raise ValueError(f"Failed to create dataset {name}: {resp.status} {resp.get_json()}")
+            raise ValueError(f"Failed to create dataset {dataset_name}: {resp.status} {resp.get_json()}")
         response_data = resp.get_json()
         dataset_id = response_data["data"]["id"]
         if dataset_id is None or dataset_id == "":
             raise ValueError(f"unexpected dataset state, invalid ID (is None: {dataset_id is None})")
         curr_version = response_data["data"]["attributes"]["current_version"]
-        return Dataset(name, dataset_id, [], description, curr_version, _dne_client=self)
+        return Dataset(dataset_name, dataset_id, [], description, curr_version, _dne_client=self)
```
Comment (sabrenner) [inline ddtrace/llmobs/_writer.py]: should/does `Dataset` need to know about `project_id` or `project_name`? even just as an attribute visible to the user

File: ddtrace/llmobs/_writer.py [hunk 5]
```diff
@@ -544,7 +561,12 @@ def project_create_or_get(self, name: str) -> str:
         if resp.status != 200:
             raise ValueError(f"Failed to create project {name}: {resp.status} {resp.get_json()}")
         response_data = resp.get_json()
-        return response_data["data"]["id"]
+        project_id = response_data["data"]["id"]
+
+        if project_id is None or project_id == "":
```
Comment (sabrenner) [inline ddtrace/llmobs/_writer.py]:
```suggestion
        if not project_id:
```

File: ddtrace/llmobs/_writer.py [hunk 6]
```diff
@@ -544,7 +561,12 @@ def project_create_or_get(self, name: str) -> str:
...
+            raise ValueError(f"project ID is required for dataset & experiments features (project name: {name})")
```
Comment (sabrenner) [inline ddtrace/llmobs/_writer.py]: more on a ux note for the sdk - are there any actionable steps we can provide here for the user to take if the provide a project name that still hits this case of no associated project id?

File: ddtrace/llmobs/_writer.py [hunk 7]
```diff
@@ -436,16 +446,19 @@ def dataset_batch_update(
...
-        path = f"/api/unstable/llm-obs/v1/datasets?filter[name]={quote(name)}"
+    def dataset_get_with_records(self, dataset_name: str, project_name: Optional[str] = None) -> Dataset:
+        project_id = self.project_create_or_get(project_name)
+        logger.debug("getting records with project ID %s for %s", project_id, project_name)
+
+        path = f"/api/unstable/llm-obs/v1/{project_id}/datasets?filter[name]={quote(dataset_name)}"
         resp = self.request("GET", path)
         if resp.status != 200:
-            raise ValueError(f"Failed to pull dataset {name}: {resp.status}")
+            raise ValueError(f"Failed to pull dataset {dataset_name} from project {project_name}: {resp.status}")
```
Comment (sabrenner) [inline ddtrace/llmobs/_writer.py]:
```suggestion
            raise ValueError(f"Failed to pull dataset {dataset_name} from project {project_name} (id={project_id}): {resp.status}")
```
would it be useful to provide the project id here as well?

File: tests/llmobs/test_experiments.py
Comment (sabrenner) [inline tests/llmobs/test_experiments.py:1]: also - we should probs add some tests that use the `project_name` feature as well, maybe some trying to hit those edge cases if possible (empty or None, no associated `project_id`) (if not already planned)

Comment (gary-huang) [reply tests/llmobs/test_experiments.py:1]: Oh yes definitely the plan, the APIs weren't ready yet but I should have the tests added soon

File: ddtrace/llmobs/_llmobs.py (hunk 1 above)
Comment (gary-huang) [reply ddtrace/llmobs/_llmobs.py:650]: correct, not GA yet, i guess now is the time to rename `name` 😆

File: ddtrace/llmobs/_writer.py (hunk 4 above)
Comment (gary-huang) [reply ddtrace/llmobs/_writer.py]: hmm, that's a good point, i think it should, i will update this!

File: ddtrace/llmobs/_llmobs.py [hunk 8]
```diff
@@ -721,6 +730,9 @@ def create_dataset_from_csv(
             # Always restore the original field size limit
             csv.field_size_limit(original_field_size_limit)
 
+        ds = cls._instance._dne_client.dataset_create(dataset_name, project_name, description)
```
Comment (gary-huang) [inline ddtrace/llmobs/_llmobs.py:733]: this is being done because local validations should take place before creating the dataset, otherwise the dataset will be created if local validations fail after, it will be annoying to work with, as the API will return 400 for trying to recreate an existing dataset

File: ddtrace/llmobs/_writer.py (hunk 6 above)
Comment (gary-huang) [reply ddtrace/llmobs/_writer.py]: 👍 logged something instead

File: ddtrace/llmobs/_writer.py (hunk 4 above)
Comment (gary-huang) [reply ddtrace/llmobs/_writer.py]: updated this to have a project class and dataset has a ref to it 👍

File: ddtrace/llmobs/_llmobs.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply ddtrace/llmobs/_llmobs.py:650]: note that this will break any current users using named arguments (might want to give some kind of heads up to existing users)

Comment (Kyle-Verhoog) [reply ddtrace/llmobs/_llmobs.py:650]: might also be worth deferring the breaking changes until the GA release so users upgrade in one go rather than breaking them per release

File: ddtrace/llmobs/_writer.py [hunk 9]
```diff
@@ -141,6 +142,7 @@ def __init__(
         _api_key: str = "",
         _app_key: str = "",
         _override_url: str = "",
+        _default_project: Project = Project(name="", _id=""),
```
Comment (Kyle-Verhoog) [inline ddtrace/llmobs/_writer.py:145]: this default is duplicated between here and the `_llmobs.py` usage. Probably better to just opt for one of them (likely just this one)

File: ddtrace/llmobs/_writer.py [hunk 10]
```diff
@@ -469,7 +487,8 @@ def dataset_get_with_records(self, name: str) -> Dataset:
             resp = self.request("GET", list_path, timeout=self.LIST_RECORDS_TIMEOUT)
             if resp.status != 200:
                 raise ValueError(
-                    f"Failed to pull {page_num}th page of dataset records {name}: {resp.status} {resp.get_json()}"
+                    f"Failed to pull {page_num}th page of dataset records {dataset_name}: "
```
Comment (Kyle-Verhoog) [inline ddtrace/llmobs/_writer.py]: > 1th

File: ddtrace/llmobs/_writer.py [hunk 11]
```diff
@@ -543,17 +564,39 @@ def dataset_bulk_upload(self, dataset_id: str, records: List[DatasetRecord]):
             raise ValueError(f"Failed to upload dataset from file: {resp.status} {resp.get_json()}")
         logger.debug("successfully uploaded with code %d", resp.status)
 
-    def project_create_or_get(self, name: str) -> str:
+    def project_create_or_get(self, name: Optional[str] = None) -> Project:
+        default_project_name = self._default_project.get("name", "")
```
Comment (Kyle-Verhoog) [inline ddtrace/llmobs/_writer.py]: nit: a little weird to have a default value in the `.get()` on the default project since there has to be a name according to the type. It also becomes a 3th definition of the the default project name

File: ddtrace/llmobs/_writer.py (hunk 10 above)
Comment (gary-huang) [reply ddtrace/llmobs/_writer.py]: 🙄

File: ddtrace/llmobs/_writer.py (hunk 9 above)
Comment (gary-huang) [reply ddtrace/llmobs/_writer.py:145]: hmmm this default is more so for the other subclasses of the base writer though, the one in _llmobs.py has an actual override on the name

File: ddtrace/llmobs/_writer.py (hunk 11 above)
Comment (gary-huang) [reply ddtrace/llmobs/_writer.py]: ok then no `get` at all

File: ddtrace/llmobs/_writer.py (hunk 9 above)
Comment (Kyle-Verhoog) [reply ddtrace/llmobs/_writer.py:145]: that's true i think that's ok then. Related, it might be worth thinking about having a `LocalProject` type which doesn't have an id field and then `Project` which does. It helps with the readability of using projects and knowing when a project is actually sync'd vs not

Comment (gary-huang) [reply ddtrace/llmobs/_writer.py:145]: yea, i think we could do `LocalProject` once projects has more fields / user interactions

File: ddtrace/llmobs/_llmobs.py (hunk 1 above)
Comment (gary-huang) [reply ddtrace/llmobs/_llmobs.py:650]: yea, this or the next minor release is going to be GA, so gonna do it now, will have the PM alert users about the changes
